crates/hyperdrive-math/src/long/targeted.rs

```diff
@@ -0,0 +1,503 @@
+use ethers::types::I256;
+use eyre::{eyre, Result};
+use fixed_point::FixedPoint;
+use fixed_point_macros::fixed;
+
+use crate::{State, YieldSpace};
+
+impl State {
+    /// Gets a target long that can be opened given a budget to achieve a desired fixed rate.
+    pub fn calculate_targeted_long_with_budget<
+        F1: Into<FixedPoint>,
+        F2: Into<FixedPoint>,
+        F3: Into<FixedPoint>,
+        I: Into<I256>,
+    >(
+        &self,
+        budget: F1,
+        target_rate: F2,
+        checkpoint_exposure: I,
+        maybe_max_iterations: Option<usize>,
+        maybe_allowable_error: Option<F3>,
+    ) -> Result<FixedPoint> {
+        let budget = budget.into();
+        match self.calculate_targeted_long(
+            target_rate,
+            checkpoint_exposure,
+            maybe_max_iterations,
+            maybe_allowable_error,
+        ) {
+            Ok(long_amount) => Ok(long_amount.min(budget)),
+            Err(error) => Err(error),
```

Comment:
I'm not sure we want to bubble this up. Let's say that I have `10_000e18` base, and the target long can't be calculated due to solvency constraints. If the max long is `100_000e18`, it would be cumbersome to get an error back from this function and have to call `calculate_max_long` manually. 

What about calling `calculate_max_long` if we hit the error case?

Reply:
made an issue to support this with a higher-level fn: https://github.com/delvtech/hyperdrive/issues/949